File: test/org/stringtemplate/test/TestWhitespace.java
package org.stringtemplate.test;

import org.junit.Test;
import static org.junit.Assert.assertEquals;
import org.stringtemplate.STGroup;
import org.stringtemplate.ST;
import org.stringtemplate.STErrorListener;
import org.stringtemplate.AutoIndentWriter;

import java.io.StringWriter;

public class TestWhitespace extends BaseTest {
    @Test public void testTrimmedSubtemplates() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<names:{n | <n>}>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = "TerTomSumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testTrimJustOneWSInSubtemplates() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<names:{n |  <n> }>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = " Ter  Tom  Sumana !";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testTrimNewlineInSubtemplates() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<names:{n |\n" +
                                     "<n>}>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = "TerTomSumana!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testLeaveNewlineOnEndInSubtemplates() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("test", "<names:{n |\n" +
                                     "<n>\n" +
                                     "}>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        String expected = "Ter\nTom\nSumana\n!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testEmptyExprAsFirstLineGetsNoOutput() throws Exception {
        ST t = new ST(
            "<users>\n" +
            "end\n");
        String expecting="end\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testEmptyLineWithIndent() throws Exception {
        ST t = new ST(
            "begin\n" +
            "    \n" +
            "end\n");
        String expecting="begin\n\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testSizeZeroOnLineByItselfGetsNoOutput() throws Exception {
        ST t = new ST(
            "begin\n"+
            "<name>\n"+
            "<users>\n"+
            "<users>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testSizeZeroOnLineWithIndentGetsNoOutput() throws Exception {
        ST t = new ST(
            "begin\n"+
            "  <name>\n"+
            "	<users>\n"+
            "	<users>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testSizeZeroOnLineWithMultipleExpr() throws Exception {
        ST t = new ST(
            "begin\n"+
            "  <name>\n"+
            "	<users><users>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testIFExpr() throws Exception {
        ST t = new ST(
            "begin\n"+
            "<if(x)><endif>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testIndentedIFExpr() throws Exception {
        ST t = new ST(
            "begin\n"+
            "    <if(x)><endif>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testIFElseExpr() throws Exception {
        ST t = new ST(
            "begin\n"+
            "<if(users)><else><endif>\n"+
            "end\n");
        String expecting="begin\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testIFOnMultipleLines() throws Exception {
        ST t = new ST(
            "begin\n"+
            "<if(users)>\n" +
            "foo\n" +
            "<else>\n" +
            "bar\n" +
            "<endif>\n"+
            "end\n");
        String expecting="begin\nbar\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testNestedIFOnMultipleLines() throws Exception {
        ST t = new ST(
            "begin\n"+
            "<if(x)>\n" +
            "<if(y)>\n" +
            "foo\n" +
            "<else>\n" +
            "bar\n" +
            "<endif>\n"+
            "<endif>\n"+
            "end\n");
        t.add("x", "x");
        String expecting="begin\nbar\nend\n";
        String result = t.render();
        assertEquals(expecting, result);
    }

    @Test public void testLineBreak() throws Exception {
        ST st = new ST(
                "Foo <\\\\>"+newline+
                "  \t  bar" +newline
                );
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\n")); // force \n as newline
        String result = sw.toString();
        String expecting ="Foo bar\n";     // expect \n in output
        assertEquals(expecting, result);
    }

    @Test public void testLineBreak2() throws Exception {
        ST st = new ST(
                "Foo <\\\\>       "+newline+
                "  \t  bar" +newline
                );
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\n")); // force \n as newline
        String result = sw.toString();
        String expecting ="Foo bar\n";
        assertEquals(expecting, result);
    }

    @Test public void testLineBreakNoWhiteSpace() throws Exception {
        ST st = new ST(
                "Foo <\\\\>"+newline+
                "bar\n"
                );
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\n")); // force \n as newline
        String result = sw.toString();
        String expecting ="Foo bar\n";
        assertEquals(expecting, result);
    }

    @Test public void testNewlineNormalizationInTemplateString() throws Exception {
        ST st = new ST(
                "Foo\r\n"+
                "Bar\n"
                );
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\n")); // force \n as newline
        String result = sw.toString();
        String expecting ="Foo\nBar\n";     // expect \n in output
        assertEquals(expecting, result);
    }

    @Test public void testNewlineNormalizationInTemplateStringPC() throws Exception {
        ST st = new ST(
                "Foo\r\n"+
                "Bar\n"
                );
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\r\n")); // force \r\n as newline
        String result = sw.toString();
        String expecting ="Foo\r\nBar\r\n";     // expect \r\n in output
        assertEquals(expecting, result);
    }

    @Test public void testNewlineNormalizationInAttribute() throws Exception {
        ST st = new ST(
                "Foo\r\n"+
                "<name>\n"
                );
        st.add("name", "a\nb\r\nc");
        StringWriter sw = new StringWriter();
        st.write(new AutoIndentWriter(sw,"\n")); // force \n as newline
        String result = sw.toString();
        String expecting ="Foo\na\nb\nc\n";     // expect \n in output
        assertEquals(expecting, result);
    }
}

File: objc/main/tests/objc/TestWhitespace.m

#import <Foundation/Foundation.h>
#import <ANTLR/ANTLR.h>
#import "AutoIndentWriter.h"
#import "STGroup.h"
#import "ST.h"
#import "CompiledST.h"
#import "StringWriter.h"
#import "TestWhitespace.h"

#define STAssertTrue GHAssertTrue

@implementation TestWhitespace

- (void) test01TrimmedSubtemplates
{
    STGroup *group = [[STGroup newSTGroup] retain];
    [group defineTemplate:@"test" argsS:@"names" template:@"<names:{n | <n>}>!"];
    ST *st = [[group getInstanceOf:@"test"] retain];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @"TerTomSumana!";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test02TrimmedSubtemplatesNoArgs
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" template:@"[<foo({ foo })>]"];
    [group defineTemplate:@"foo" argsS:@"x" template:@"<x>"];
    ST *st = [group getInstanceOf:@"test"];
    NSString *expected = @"[ foo ]";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test03TrimmedSubtemplatesArgs
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" argsS:@"names" template:@"<names:{x|  foo }>"];
    ST *st = [group getInstanceOf:@"test"];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @" foo  foo  foo ";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test04TrimJustOneWSInSubtemplates
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" argsS:@"names" template:@"<names:{n |  <n> }>!"];
    ST *st = [group getInstanceOf:@"test"];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @" Ter  Tom  Sumana !";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test05TrimNewlineInSubtemplates
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" argsS:@"names" template:@"<names:{n |\n<n>}>!"];
    ST *st = [group getInstanceOf:@"test"];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @"TerTomSumana!";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test06LeaveNewlineOnEndInSubtemplates
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" argsS:@"names" template:@"<names:{n |\n<n>\n}>!"];
    ST *st = [group getInstanceOf:@"test"];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @"Ter\nTom\nSumana\n!";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test07TabBeforeEndInSubtemplates
{
    STGroup *group = [STGroup newSTGroup];
    [group defineTemplate:@"test" argsS:@"names" template:@"  <names:{n |\n    <n>\n  }>!"];
    ST *st = [group getInstanceOf:@"test"];
    [st add:@"names" value:@"Ter"];
    [st add:@"names" value:@"Tom"];
    [st add:@"names" value:@"Sumana"];
    NSString *expected = @"      Ter\n      Tom\n      Sumana\n!";
    NSString *result = [st render];
    [st.impl dump];
    NSLog( @"expected string length =%d, result string length = %d\n", [expected length], [result length] );
    [self assertEquals:expected result:result];
    return;
}

- (void) test08EmptyExprAsFirstLineGetsNoOutput
{
    ST *st = [ST newSTWithTemplate:@"<users>\nend\n"];
    NSString *expected = @"end\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test09EmptyLineWithIndent
{
    ST *st = [ST newSTWithTemplate:@"begin\n    \nend\n"];
    NSString *expected = @"begin\n\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test10EmptyLine
{
    ST *st = [ST newSTWithTemplate:@"begin\n\nend\n"];
    NSString *expected = @"begin\n\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test11SizeZeroOnLineByItselfGetsNoOutput
{
    ST *st = [ST newSTWithTemplate:@"begin\n<name>\n<users>\n<users>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test12SizeZeroOnLineWithIndentGetsNoOutput
{
    ST *st = [ST newSTWithTemplate:@"begin\n  <name>\n	<users>\n	<users>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test13SizeZeroOnLineWithMultipleExpr
{
    ST *st = [ST newSTWithTemplate:@"begin\n  <name>\n	<users><users>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test14IFExpr
{
    ST *st = [ST newSTWithTemplate:@"begin\n<if(x)><endif>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test15IndentedIFExpr
{
    ST *st = [ST newSTWithTemplate:@"begin\n    <if(x)><endif>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test16IFElseExpr
{
    ST *st = [ST newSTWithTemplate:@"begin\n<if(users)><else><endif>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test17IFOnMultipleLines
{
    ST *st = [ST newSTWithTemplate:@"begin\n<if(users)>\nfoo\n<else>\nbar\n<endif>\nend\n"];
    NSString *expected = @"begin\nbar\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test18NestedIFOnMultipleLines
{
    ST *st = [ST newSTWithTemplate:@"begin\n<if(x)>\n<if(y)>\nfoo\n<else>\nbar\n<endif>\n<endif>\nend\n"];
    [st add:@"x" value:@"x"];
    NSString *expected = @"begin\nbar\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test19LineBreak
{
    ST *st = [ST newSTWithTemplate:@"Foo <\\\\>\n  \t  bar\n"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\n"]];
    NSString *result = [sw description];
    NSString *expected = @"Foo bar\n";
    [self assertEquals:expected result:result];
    return;
}

- (void) test20LineBreak2
{
    ST *st = [ST newSTWithTemplate:@"Foo <\\\\>       \n  \t  bar\n"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\n"]];
    NSString *result = [sw description];
    NSString *expected = @"Foo bar\n";
    [self assertEquals:expected result:result];
    return;
}

- (void) test21LineBreakNoWhiteSpace
{
    ST *st = [ST newSTWithTemplate:@"Foo <\\\\>\nbar\n"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\n"]];
    NSString *result = [sw description];
    NSString *expected = @"Foo bar\n";
    [self assertEquals:expected result:result];
    return;
}

- (void) test22NewlineNormalizationInTemplateString
{
    ST *st = [ST newSTWithTemplate:@"Foo\r\nBar\n"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\n"]];
    NSString *expected = @"Foo\nBar\n";
    NSString *result = [sw description];
    [self assertEquals:expected result:result];
    return;
}

- (void) test23NewlineNormalizationInTemplateStringPC
{
    ST *st = [ST newSTWithTemplate:@"Foo\r\nBar\n"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\r\n"]];
    NSString *expected = @"Foo\r\nBar\r\n";
    NSString *result = [sw description];
    [self assertEquals:expected result:result];
    return;
}

- (void) test24NewlineNormalizationInAttribute
{
    ST *st = [ST newSTWithTemplate:@"Foo\r\n<name>\n"];
    [st add:@"name" value:@"a\nb\r\nc"];
    StringWriter *sw = [StringWriter newWriter];
    [st write:[AutoIndentWriter newWriter:sw newLine:@"\n"]];
    NSString *expected = @"Foo\na\nb\nc\n";
    NSString *result = [sw description];
    [self assertEquals:expected result:result];
    return;
}

- (void) test25CommentOnlyLineGivesNoOutput
{
    ST *st = [ST newSTWithTemplate:@"begin\n<! ignore !>\nend\n"];
    NSString *expected = [NSString stringWithFormat:@"begin%@end%@", @"\n",  @"\n" ];
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

- (void) test26CommentOnlyLineGivesNoOutput2
{
    ST *st = [ST newSTWithTemplate:@"begin\n    <! ignore !>\nend\n"];
    NSString *expected = @"begin\nend\n";
    NSString *result = [st render];
    [self assertEquals:expected result:result];
    return;
}

@end
